components/registration-framework/org.wso2.carbon.identity.user.registration.engine/src/main/java/org/wso2/carbon/identity/user/registration/engine/util/RegistrationFlowEngineUtils.java

```diff
@@ -206,4 +230,53 @@ public static String buildMyAccountAccessURL(String tenantDomain) {
 
         return ApplicationMgtUtil.getMyAccountAccessUrlFromServerConfig(tenantDomain);
     }
+
+    /**
+     * Resolve the registration completion redirection URL.
+     *
+     * @param context   Registration context.
+     * @return  Redirection URL.
+     * @throws RegistrationEngineServerException    Registration framework exception.
+     */
+    public static String resolveCompletionRedirectionUrl(RegistrationContext context)
+            throws RegistrationEngineServerException {
+
+        String redirectionUrl = getApplicationAccessUrl(context.getTenantDomain(), context.getApplicationId());
+
+        // If the application access URL is not available, we will use the default callback URL.
+        if (StringUtils.isEmpty(redirectionUrl)) {
+            redirectionUrl = context.getCallbackUrl();
+        }
+
+        // If the callback URL is not available, we will use the MyAccount access URL.
+        if (StringUtils.isEmpty(redirectionUrl)) {
+            redirectionUrl = buildMyAccountAccessURL(context.getTenantDomain());
+        }
+        return redirectionUrl;
+    }
+
+    /**
+     * Get the application access URL.
+     *
+     * @param tenantDomain  Tenant domain.
+     * @param applicationId Application ID.
+     * @return Application access URL.
+     * @throws RegistrationEngineServerException Registration framework exception.
+     */
+    private static String getApplicationAccessUrl(String tenantDomain, String applicationId)
+            throws RegistrationEngineServerException {
+
+        ServiceProvider application;
+        ApplicationManagementService applicationManagementService =
+                RegistrationFlowEngineDataHolder.getInstance().getApplicationManagementService();
+        try {
+            application = applicationManagementService.getApplicationByResourceId(applicationId, tenantDomain);
```

Comment:
is it possible to obtain the access URL from application basic info rather than fetching the complete service provider?

Reply:
There's no existing service in the ApplicationManagementService to retrieve the application basic info using the application id.